db/db_iter.cc

```diff
@@ -611,21 +613,27 @@ inline bool DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check)
             !iter_.iter()->IsKeyPinned() || !pin_thru_lifetime_ /* copy */);
         skipping = false;
         num_skipped = 0;
+        reseek_done = false;
       }
     }
 
     // If we have sequentially iterated via numerous equal keys, then it's
     // better to seek so that we can avoid too many key comparisons.
-    if (num_skipped > max_skip_ && CanReseekToSkip()) {
+    //
+    // Do not reseek if we have already attempted to reseek previously to avoid
+    // infinite loops.
+    if (num_skipped > max_skip_ && !reseek_done) {
```

Comment:
Talked offline to @maysamyabandeh, and he suggested we could optimize this further by checking if the seek sequence number is greater than the current key sequence number. If it's greater, than there's no point reseeking because we would actually set ourselves back, and we should just continue iterating instead.